File: squelch-desktop/src/lib/threadPrefetch.ts
// Thread prefetch + image warmer — the "instant open" machinery.
//
// The thread viewer used to cold-start on every open: fetch the thread, build
// the srcdoc, then let images trickle in over the network. This module lets the
// inbox warm all of that BEFORE the click:
//   - prefetchThread(id): fetch + LRU-cache the ClientThreadView, then warm
//     every image it references into lib/imageCache (Rust-side fetch, held as
//     data: URIs) so the frame can be built with the bytes already in hand.
//   - getPrefetchedThread(id): a fresh cached view (or null) — the viewer
//     renders it synchronously on mount, no loading flash, no refetch.
//
// WARMING WAS REWRITTEN (2026-07). It used to create parent-document Image()
// objects and hope the shared WKWebView HTTP cache held them — capped at 12
// <img src> per message and blind to CSS url(). Both holes are why image-heavy
// mail still flickered on open; see the header of lib/imageCache for the full
// account. Warming now covers EVERY reference in EVERY message of the thread,
// and the bytes are pinned rather than hoped for. The per-message cap is gone:
// lib/imageCache's global byte budget is the real limiter, so coverage is
// bounded by memory rather than by an arbitrary count that silently truncated
// exactly the below-the-fold images that do the reflowing.
//
// PRIVACY: warming loads remote images for mail the user hasn't opened, so it
// (a) strips tracking pixels first (same stripTrackers pass the frame uses),
// (b) never runs when the Settings "load on demand" images pref is on, and
// (c) sends no referrer (the shell fetch carries no cookies or Referer).

import { api } from "../api";
import type { ClientThreadView } from "../api";
import { getPrefs } from "./prefs";
import { stripTrackers } from "./trackers";
import { warmImageCache } from "./imageCache";

/* Sized to hold the whole For-your-eyes list (sitrep preloads every standing
   item) plus inbox hover-warms without LRU churn. Views are small (text/html
   strings); 60 is a few MB worst-case. */
const CACHE_MAX = 60;
/** A cached view older than this refetches on real open (mail can change). */
const FRESH_MS = 60_000;
const cache = new Map<
  string,
  { view: ClientThreadView; ts: number; freshMs: number }
>();
const inflight = new Set<string>();
/** Promise-shaped in-flight dedupe for callers that need the VIEW back
 *  (newsletter hero thumbnails) rather than fire-and-forget warming. */
const inflightPromises = new Map<string, Promise<ClientThreadView>>();

/**
 * Pin every image in every message of a view (pref-gated). Fire-and-forget:
 * lib/imageCache dedupes by URL, negative-caches failures and bounds its own
 * concurrency, so re-warming an already-warm thread costs a map lookup per
 * reference. Warming is deliberately NOT urgent — an email actually being
 * opened promotes its own fetches ahead of this background sweep.
 */
function warmImages(view: ClientThreadView): void {
  if (!getPrefs().loadRemoteImages) return;
  for (const msg of view.messages) {
    if (!msg.html) continue;
    const { html } = stripTrackers(msg.html);
    void warmImageCache(html);
  }
}

function put(threadId: string, view: ClientThreadView, freshMs: number): void {
  cache.delete(threadId); // re-insert => most-recent position
  cache.set(threadId, { view, ts: Date.now(), freshMs });
  if (cache.size > CACHE_MAX) {
    const oldest = cache.keys().next().value;
    if (oldest !== undefined) cache.delete(oldest);
  }
}

/** Fire-and-forget: fetch + cache + warm. Deduped while in flight; a fresh
 *  cache hit just re-warms (Image() from HTTP cache is ~free). */
export function prefetchThread(
  threadId: string,
  opts?: { freshMs?: number },
): void {
  // Per-entry TTL: right-rail records (banking/receipts) stay valid as long as
  // their column shows them, so their cached threads outlive the 60s default.
  // A repeat prefetch may EXTEND an entry's ttl (never shorten it).
  const freshMs = opts?.freshMs ?? FRESH_MS;
  const hit = cache.get(threadId);
  if (hit && Date.now() - hit.ts < Math.max(hit.freshMs, freshMs)) {
    if (freshMs > hit.freshMs) hit.freshMs = freshMs;
    warmImages(hit.view);
    return;
  }
  if (inflight.has(threadId)) return;
  inflight.add(threadId);
  api
    .getThread(threadId)
    .then((view) => {
      put(threadId, view, freshMs);
      warmImages(view);
    })
    .catch(() => {
      // Prefetch is best-effort; the real open will surface any error.
    })
    .finally(() => inflight.delete(threadId));
}

/** A fresh cached view for instant render, or null. */
export function getPrefetchedThread(threadId: string): ClientThreadView | null {
  const hit = cache.get(threadId);
  if (!hit || Date.now() - hit.ts >= hit.freshMs) return null;
  return hit.view;
}

/**
 * Fetch a thread THROUGH the cache, returning the view: a fresh cache hit
 * resolves immediately; concurrent callers share one request. Used by the
 * newsletter hero thumbnails (they need the html, not just warming).
 */
export function fetchThreadCached(
  threadId: string,
  opts?: { freshMs?: number },
): Promise<ClientThreadView> {
  const freshMs = opts?.freshMs ?? FRESH_MS;
  const hit = cache.get(threadId);
  if (hit && Date.now() - hit.ts < Math.max(hit.freshMs, freshMs)) {
    if (freshMs > hit.freshMs) hit.freshMs = freshMs;
    return Promise.resolve(hit.view);
  }
  const pending = inflightPromises.get(threadId);
  if (pending) return pending;
  const p = api
    .getThread(threadId)
    .then((view) => {
      put(threadId, view, freshMs);
      warmImages(view);
      return view;
    })
    .finally(() => inflightPromises.delete(threadId));
  inflightPromises.set(threadId, p);
  return p;
}

/** Let the viewer's own (authoritative) fetch feed the cache + warm images —
 *  the next reopen of the same thread is then instant too. */
export function noteFetchedThread(threadId: string, view: ClientThreadView): void {
  put(threadId, view, FRESH_MS);
  warmImages(view);
}

File: squelch-client-swift/Sources/Squelch/Model/ThreadPrefetch.swift
// Thread prefetch — the "instant open" machinery.
//
// The thread viewer used to cold-start on every open: fetch the thread, build
// the document, then let images trickle in. This lets the inbox warm all of it
// BEFORE the click:
//   - prefetch(id): fetch + LRU-cache the ClientThreadView.
//   - cached(id):   a fresh cached view (or nil) — the viewer renders it
//                   synchronously on mount, no loading flash, no refetch.
//
// The WKWebView handles image loading itself (and its shared URL cache means a
// warmed thread's images are usually already resident), so unlike the Tauri
// build there is no separate byte-pinning cache to maintain — the flicker that
// forced one came from srcdoc rebuilds, which don't happen here.
//
// Caching the JSON is only half of an instant open, though: every body still
// had to be tracker-stripped, de-duped and link-scanned before it could be
// handed to a frame, and that ran on the main actor at open time. So a cached
// thread ALSO warms its prepared bodies (PreparedBodies) off the main actor —
// see warmBodies.
//
// Also holds the measured-height memory, so reopening a message renders at its
// final size on the first paint.

import Foundation

@MainActor
final class ThreadPrefetch {
    static let shared = ThreadPrefetch()

    /// Sized to hold the whole For-your-eyes list (the sitrep preloads every
    /// standing item) plus inbox hover-warms without LRU churn.
    private static let cacheMax = 60
    /// A cached view older than this refetches on real open (mail can change).
    private static let freshDefault: TimeInterval = 60

    private struct Entry {
        var view: ClientThreadView
        var ts: Date
        var fresh: TimeInterval
    }

    private var cache: [String: Entry] = [:]
    /// Insertion order for LRU eviction (most-recent last).
    private var order: [String] = []
    private var inflight: Set<String> = []
    /// threadId -> the warmer's repeated-image map. Rides the same LRU as the
    /// views it was derived from, because it is only meaningful next to one.
    private var repeated: [String: [Int: Set<String>]] = [:]

    private init() {}

    private func put(_ threadId: String, _ view: ClientThreadView, fresh: TimeInterval) {
        cache[threadId] = Entry(view: view, ts: Date(), fresh: fresh)
        order.removeAll { $0 == threadId }
        order.append(threadId)
        while order.count > Self.cacheMax {
            let oldest = order.removeFirst()
            cache.removeValue(forKey: oldest)
            repeated.removeValue(forKey: oldest)
        }
        warmBodies(threadId, view)
    }

    /// Fire-and-forget: fetch + cache. Deduped while in flight; a fresh cache
    /// hit is a no-op.
    ///
    /// `fresh` is a per-entry TTL: right-rail records (banking/receipts) stay
    /// valid as long as their column shows them, so their cached threads
    /// outlive the 60s default. A repeat prefetch may EXTEND a TTL, never
    /// shorten it.
    func prefetch(_ threadId: String, fresh: TimeInterval? = nil) {
        guard !threadId.isEmpty else { return }
        let ttl = fresh ?? Self.freshDefault
        if var hit = cache[threadId], Date().timeIntervalSince(hit.ts) < max(hit.fresh, ttl) {
            if ttl > hit.fresh {
                hit.fresh = ttl
                cache[threadId] = hit
            }
            return
        }
        guard !inflight.contains(threadId) else { return }
        inflight.insert(threadId)
        Task { [weak self] in
            defer { self?.inflight.remove(threadId) }
            // Prefetch is best-effort; the real open surfaces any error.
            guard let view = try? await APIClient.shared.getThread(threadId) else { return }
            self?.put(threadId, view, fresh: ttl)
        }
    }

    /// A fresh cached view for instant render, or nil.
    func cached(_ threadId: String) -> ClientThreadView? {
        guard let hit = cache[threadId], Date().timeIntervalSince(hit.ts) < hit.fresh else {
            return nil
        }
        return hit.view
    }

    /// Fetch THROUGH the cache, returning the view: a fresh hit resolves
    /// immediately. Used by newsletter hero thumbnails, which need the html.
    func fetch(_ threadId: String, fresh: TimeInterval? = nil) async throws -> ClientThreadView {
        let ttl = fresh ?? Self.freshDefault
        if let hit = cache[threadId], Date().timeIntervalSince(hit.ts) < max(hit.fresh, ttl) {
            return hit.view
        }
        let view = try await APIClient.shared.getThread(threadId)
        put(threadId, view, fresh: ttl)
        return view
    }

    /// Let the viewer's own (authoritative) fetch feed the cache — the next
    /// reopen of the same thread is then instant too.
    func note(_ threadId: String, _ view: ClientThreadView) {
        put(threadId, view, fresh: Self.freshDefault)
    }

    // MARK: - prepared bodies

    /// The warmer's repeated-image map for a thread, or nil if it has not
    /// finished (or the thread was evicted). nil is a normal answer — the
    /// viewer computes its own rather than waiting on us.
    func cachedRepeatedImages(_ threadId: String) -> [Int: Set<String>]? { repeated[threadId] }

    /// Preprocess every body in a freshly cached thread OFF the main actor.
    ///
    /// These are the exact scans EmailWebView used to run in a `.task` at open
    /// time, per message, on the main actor — so opening a thread paid for a
    /// full regex walk of every body in it PLUS a runloop beat before the first
    /// frame had any html to load. Running them here means the cache that
    /// already makes the open network-free makes it scan-free too.
    ///
    /// Fire-and-forget and idempotent: a key already in PreparedBodies is
    /// skipped, so re-prefetching a thread costs one dictionary probe per
    /// message. Everything crossing out is a value type (ClientThreadView,
    /// Prepared) and Trackers/ImageRepeats are pure, so nothing here needs the
    /// main actor except the hand-back at the end.
    private func warmBodies(_ threadId: String, _ view: ClientThreadView) {
        Task.detached(priority: .utility) {
            let map = Self.repeatedImages(in: view)
            for message in view.messages {
                guard let html = message.html, !html.isEmpty else { continue }
                let seenEarlier = map[message.id] ?? []
                let key = EmailWebView.Prepared.cacheKey(html, seenEarlier)
                guard PreparedBodies.shared.get(key) == nil else { continue }
                PreparedBodies.shared.set(
                    key, EmailWebView.Prepared.make(from: html, seenEarlier: seenEarlier))
            }
            await MainActor.run { ThreadPrefetch.shared.noteRepeated(threadId, map) }
        }
    }

    /// Only remember the map while the view it describes is still cached — a
    /// thread evicted mid-scan must not leave its map behind.
    private func noteRepeated(_ threadId: String, _ map: [Int: Set<String>]) {
        guard cache[threadId] != nil else { return }
        repeated[threadId] = map
    }

    /// messageId -> image srcs already shown by an earlier message.
    ///
    /// "EARLIER" MEANS CHRONOLOGICALLY EARLIER, so this walks `view.messages`
    /// (server order, oldest first) and NOT the newest-first order the reader
    /// sees. Walking the display order instead would keep the newest copy of a
    /// signature and suppress the original — hiding the image in the message
    /// that actually introduced it.
    ///
    /// Scans TRACKER-STRIPPED html for the same reason the strip runs first in
    /// EmailWebView: a tracking pixel is removed from every message anyway, so
    /// letting one register as a "first occurrence" here could suppress a real
    /// image that shares its src and leave the thread showing none at all.
    ///
    /// `nonisolated` because the warmer runs it off the main actor, and the
    /// viewer's cold path runs it on.
    nonisolated static func repeatedImages(in view: ClientThreadView) -> [Int: Set<String>] {
        var seen = Set<String>()
        var out: [Int: Set<String>] = [:]
        for message in view.messages {
            out[message.id] = seen
            guard let html = message.html, !html.isEmpty else { continue }
            seen.formUnion(ImageRepeats.sources(Trackers.strip(html).html))
        }
        return out
    }

    // MARK: - batch warming

    /// Warm a staggered batch so a fresh list never stampedes the daemon.
    /// `immediate` rows go now; the rest trickle at `spacing`.
    func warm(_ threadIds: [String], immediate: Int = 5, spacing: Duration = .milliseconds(120)) {
        for id in threadIds.prefix(immediate) { prefetch(id) }
        let rest = Array(threadIds.dropFirst(immediate))
        guard !rest.isEmpty else { return }
        Task { [weak self] in
            for id in rest {
                try? await Task.sleep(for: spacing)
                if Task.isCancelled { return }
                self?.prefetch(id)
            }
        }
    }
}

/// Preprocessed email bodies, keyed by `EmailWebView.Prepared.cacheKey` — the
/// html AND the suppression set that produced them, because the same body
/// prepared against a different thread is a different document.
///
/// LOCK-BASED RATHER THAN @MainActor, and that is the entire point of it: the
/// warmer fills it from a detached task while `EmailWebView.init` — which
/// SwiftUI runs nonisolated, before any `.task` or `onAppear` can — reads it to
/// seed its state. An actor-isolated cache could serve neither of those without
/// an await, and an await in a View initializer is not a thing.
final class PreparedBodies: @unchecked Sendable {
    static let shared = PreparedBodies()

    /// Several threads' worth of messages: enough that walking a queue back and
    /// forth never re-scans, bounded because each entry holds a copy of a
    /// message body.
    private static let cap = 300

    private let lock = NSLock()
    private var entries: [Int: EmailWebView.Prepared] = [:]
    /// Insertion order for LRU eviction (most-recent last).
    private var order: [Int] = []

    private init() {}

    func get(_ key: Int) -> EmailWebView.Prepared? {
        lock.lock()
        defer { lock.unlock() }
        return entries[key]
    }

    func set(_ key: Int, _ prepared: EmailWebView.Prepared) {
        lock.lock()
        defer { lock.unlock() }
        if entries.updateValue(prepared, forKey: key) == nil { order.append(key) }
        while order.count > Self.cap {
            entries.removeValue(forKey: order.removeFirst())
        }
    }
}

/// Remembered rendered heights for email bodies, keyed by message id. A frame
/// with a remembered height renders at its final size instantly — zero resize
/// on reopen, which is what the reader perceives as flicker.
@MainActor
final class FrameHeights {
    static let shared = FrameHeights()
    private var heights: [String: CGFloat] = [:]
    private init() {}

    func get(_ key: String) -> CGFloat? { heights[key] }
    func set(_ key: String, _ height: CGFloat) { heights[key] = height }
    func clear(_ key: String) { heights.removeValue(forKey: key) }
}
